File: src/shared/api/ws/policy.ts
export type PolicySnapshot = {
  policyVersion: string
  expiresAtMs: number
  verified: boolean
  killSwitch: 'clear' | 'stop_new' | 'stop_all'
}

type PolicyControllerDependencies = {
  stopAll: (reason: 'policy_lost' | 'kill_switch') => void | Promise<void>
  cancelPending: () => void
  emitAudit: (reason: 'POLICY_LOST' | 'POLICY_STALE' | 'KILL_SWITCH_ACTIVE') => void
  now?: () => number
}

export function createPolicyController(dependencies: PolicyControllerDependencies) {
  const now = dependencies.now ?? Date.now
  let snapshot: PolicySnapshot | undefined
  let stopped = false

  async function stopOnce(reason: 'policy_lost' | 'kill_switch', auditReason: 'POLICY_LOST' | 'POLICY_STALE' | 'KILL_SWITCH_ACTIVE') {
    if (stopped) return false
    stopped = true
    dependencies.cancelPending()
    try {
      await dependencies.stopAll(reason)
    } finally {
      dependencies.emitAudit(auditReason)
    }
    return true
  }

  return {
    subscribe(deviceId: string) {
      return { type: 'policy.subscribe' as const, payload: { deviceId } }
    },
    async apply(next: PolicySnapshot) {
      snapshot = next
      if (!next.verified || next.expiresAtMs <= now()) return stopOnce('policy_lost', 'POLICY_STALE')
      if (next.killSwitch !== 'clear') return stopOnce('kill_switch', 'KILL_SWITCH_ACTIVE')
      return false
    },
    async tick() {
      if (!snapshot || !snapshot.verified || snapshot.expiresAtMs <= now()) {
        return stopOnce('policy_lost', snapshot ? 'POLICY_STALE' : 'POLICY_LOST')
      }
      return false
    },
    transportLost() {
      snapshot = undefined
      return stopOnce('policy_lost', 'POLICY_LOST')
    },
    blocked: () => stopped,
  }
}

File: src-tauri/src/security/policy.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum KillSwitch {
    Clear,
    StopNew,
    StopAll,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SafetyPolicySnapshot {
    pub policy_version: String,
    pub user_id: String,
    pub device_id: String,
    pub environment_id: String,
    pub allow_adversarial: bool,
    pub allow_export: bool,
    pub kill_switch: KillSwitch,
    pub expires_at_ms: i64,
    pub verified: bool,
}

impl SafetyPolicySnapshot {
    pub fn is_fresh(&self, now_ms: i64) -> bool {
        self.verified && self.expires_at_ms > now_ms
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum PolicyAction {
    Continue,
    StopAll,
    AlreadyStopped,
}

#[derive(Debug, Default)]
pub struct PolicyEnforcer {
    snapshot: Option<SafetyPolicySnapshot>,
    stopped: bool,
}

impl PolicyEnforcer {
    pub fn apply(&mut self, snapshot: SafetyPolicySnapshot, now_ms: i64) -> PolicyAction {
        if self.stopped {
            return PolicyAction::AlreadyStopped;
        }
        let permitted = snapshot.is_fresh(now_ms) && snapshot.kill_switch == KillSwitch::Clear;
        self.snapshot = Some(snapshot);
        if permitted {
            PolicyAction::Continue
        } else {
            self.stop_once()
        }
    }

    pub fn tick(&mut self, now_ms: i64) -> PolicyAction {
        if self.snapshot.as_ref().is_some_and(|snapshot| {
            snapshot.is_fresh(now_ms) && snapshot.kill_switch == KillSwitch::Clear
        }) {
            PolicyAction::Continue
        } else {
            self.stop_once()
        }
    }

    pub fn transport_lost(&mut self) -> PolicyAction {
        self.snapshot = None;
        self.stop_once()
    }

    fn stop_once(&mut self) -> PolicyAction {
        if std::mem::replace(&mut self.stopped, true) {
            PolicyAction::AlreadyStopped
        } else {
            PolicyAction::StopAll
        }
    }
}
